HeterogeneousCore/AlpakaCore/README.md

```diff
@@ -0,0 +1,403 @@
+# Alpaka algorithms and modules in CMSSW
+
+## Introduction
+
+This page documents the Alpaka integration within CMSSW. For more information about Alpaka itself see the [Alpaka documentation](https://alpaka.readthedocs.io/en/latest/).
+
+### Compilation model
+
+The code in `Package/SubPackage/{interface,src,plugins,test}/alpaka` is compiled once for each enabled Alpaka backend. The `ALPAKA_ACCELERATOR_NAMESPACE` macro is substituted with a concrete, backend-specific namespace name in order to guarantee different symbol names for all backends, that allows for `cmsRun` to dynamically load any set of the backend libraries.
+
+The source files with `.dev.cc` suffix are compiled with the backend-specific device compiler. The other `.cc` source files are compiled with the host compiler.
+
+The `BuildFile.xml` must contain `<flags ALPAKA_BACKENDS="1"/>` to enable the behavior described above.
+
+## Overall guidelines
+
+* Minimize explicit blocking synchronization calls
+  * Avoid `alpaka::wait()`, non-cached memory buffer allocations
+* If you can, use `global::EDProducer` base class
+  * If you need per-stream storage
+    * For few objects consider using [`edm::StreamCache<T>`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_StreamCacheT) with the global module, or
+    * Use `stream::EDProducer`
+  * If you need to transfer some data back to host, use `stream::SynchronizingEDroducer`
+* All code using `ALPAKA_ACCELERATOR_NAMESPACE` should be placed in `Package/SubPackage/{interface,src,plugins,test}/alpaka` directory
+  * Alpaka-dependent code that uses templates instead of the namespace macro can be placed in `Package/SubPackage/interface` directory
+* All source files (not headers) using Alpaka device code (such as kernel call, functions called by kernels) must have a suffic `.dev.cc`, and be placed in the aforementioned `alpaka` subdirectory
+* Any code that `#include`s a header from the framework or from the `HeterogeneousCore/AlpakaCore`(?) must be separated from the Alpaka device code, and have the usual `.cc` suffix.
+  * A few framework headers are allowed to be used in `.dev.cc` files, e.g. `FWCore/Utilities/interface/Exception.h` or `FWCore/MessageLogger/interface/MessageLogger.h` (?)
+
+## Data formats
+
+Data formats, for both Event and EventSetup, should be placed following their usual rules. The Alpaka-specific conventions are
+* There must be a host-only flavor of the data format that is either idependent of Alpaka, or depends only on Alpaka's Serial backend
+  * The host-only data format must be defined in `Package/SubPackage/interface/` directory
```

Comment:
```suggestion
  * The host-only data format must be defined in `DataFormats/Package/interface/` directory
```
?

Reply:
Technically there are other data format packages than `DataFormats` (e.g. `SimDataFormats`). Fully transient data products are also allowed to be defined in packages outside of the `*DataFormats` (and with looser constraints).

I agree seeing `DataFormats` explicitly here could be helpful for e.g. newcomers. A possible compromise would be to use `DataFormats` in these placement guidelines, and after the bullet list add a link to https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts for the exact constraints for data format classes. What do you think?

Reply:
> Technically there are other data format packages than `DataFormats` (e.g. `SimDataFormats`).

True, but I don't see us making changes to the simulation DataFormats any time soon...


> Fully transient data products are also allowed to be defined in packages outside of the `*DataFormats` (and with looser constraints)

I admit I'm not familiar with the use cases for transient data formats, other than for testing and `CUDADataFormats` (and I think we want to get rid of the latter).


> A possible compromise would be ...

Sounds good to me.

Reply:
In the end I used the `DataFormats/SubPackage` only for the cases that referred to Event data products, and kept the more general `Package/SubPackage` elsewhere, as only a few EventSetup data format classes are defined in `DataFormats`.